hivemind/optim/experimental/progress_tracker.py

```diff
@@ -0,0 +1,326 @@
+        self.local_progress = self._get_local_progress(local_epoch=0, samples_accumulated=0)
+        metadata, _expiration = self.dht.get(self.training_progress_key, latest=True) or (None, -float("inf"))
+        self.global_progress = self._parse_swarm_progress_data(metadata)
+        self.lock_global_progress, self.global_state_updated = Lock(), Event()
+        self.lock_local_progress, self.should_report_progress = Lock(), Event()
+        self.shutdown_triggered, self.shutdown_complete = Event(), Event()
+        super().__init__(name=f"{self.__class__.__name__}({self.prefix})", daemon=daemon)
+        if start:
+            self.start()
+
+    @property
+    def global_epoch(self) -> int:
+        return self.global_progress.epoch
+
+    @property
+    def ready_to_increment_epoch(self) -> bool:
+        """Whether or not this peer can increment epoch right away."""
+        if self.global_epoch > self.local_progress.epoch:
+            return True
+        elif self.global_progress.samples_accumulated >= self.target_batch_size:
+            return True
+        elif get_dht_time() >= self.global_progress.eta_next_epoch:
+            return True
+        return False
+
+    @property
+    def estimated_next_increment_time(self) -> DHTExpiration:
+        """Estimate (absolute) time when this peer should increment epoch"""
+        if self.ready_to_increment_epoch:
+            return get_dht_time()
+        return self.global_progress.eta_next_epoch
+
+    def _get_local_progress(self, local_epoch: int, samples_accumulated: int):
+        return LocalTrainingProgress(
+            peer_id=self.dht.peer_id.to_bytes(),
+            epoch=local_epoch,
+            samples_accumulated=samples_accumulated,
+            samples_per_second=self.performance_ema.samples_per_second,
+            time=get_dht_time(),
+            client_mode=self.client_mode,
+        )
+
+    def report_local_progress(self, local_epoch: int, samples_accumulated: int):
+        """Update the number of locally accumulated samples and notify to other peers about this."""
+        with self.lock_local_progress:
+            extra_samples = samples_accumulated - self.local_progress.samples_accumulated
+            if extra_samples > 0:
+                self.performance_ema.update(task_size=extra_samples)
+                logger.debug(f"Updated performance EMA: {self.performance_ema.samples_per_second:.5f}")
+            else:
+                logger.debug("Resetting performance timestamp to current time (progress was reset)")
+                self.performance_ema.reset_timer()
+            self.local_progress = self._get_local_progress(local_epoch, samples_accumulated)
+            self.should_report_progress.set()
+
+    @contextlib.contextmanager
+    def pause_updates(self):
+        """Temporarily stop progress tracker from updating global training state"""
+        with self.lock_global_progress, self.performance_ema.pause():
```

Comment:
(may be out of scope of this PR)

I'm surprised that `performance_ema.pause()` resets the timer instead of actually pausing it.